File: app/models/qualification.rb
class Qualification < ApplicationRecord
  # Relationships
  belongs_to :qualifiable, polymorphic: true
  belongs_to :value_list, counter_cache: true

  # Delegates
  delegate :group, to: :value_list, prefix: true
  delegate :object, to: :value_list, prefix: true

  # Validations
  validate :validate_value_list

  private

  def validate_value_list
    if value_list_id == qualifiable_id && qualifiable_type == ValueList.to_s
      errors.add(:value_list_id, I18n.t('errors.qualification.self_reference'))
    end
  end
end

File: client/src/types/Qualification.js
// @flow

import type { ValueList } from './ValueList';

export type Qualification = {
  id: number,
  qualifiable_id: number,
  qualifiable_type: string,
  value_list_id: number,
  value_list: ValueList,
  form_field: string,
  notes: any,
  persistent: boolean
};
